libs/estdlib/src/supervisor.erl

```diff
@@ -169,75 +263,44 @@ child_spec_to_record(#{id := ChildId, start := MFA} = ChildMap) ->
     case try_start(Child) of
         {ok, Pid, _Result} ->
             start_children(T, [Child#child{pid = Pid} | StartedC])
     end;
 start_children([], StartedC) ->
-    StartedC.
-
-restart_child(Pid, Reason, State) ->
-    case lists:keyfind(Pid, #child.pid, State#state.children) of
-        false ->
-            {ok, State};
-        #child{restart = {terminating, temporary, From}} ->
-            gen_server:reply(From, ok),
-            NewChildren = lists:keydelete(Pid, #child.pid, State#state.children),
-            {ok, State#state{children = NewChildren}};
-        #child{restart = {terminating, Restart, From}} = Child ->
-            gen_server:reply(From, ok),
-            NewChildren = lists:keyreplace(Pid, #child.pid, State#state.children, Child#child{
-                pid = undefined, restart = Restart
-            }),
-            {ok, State#state{children = NewChildren}};
-        #child{} = Child ->
-            case should_restart(Reason, Child#child.restart) of
-                true ->
-                    case try_start(Child) of
-                        {ok, NewPid, _Result} ->
-                            NewChild = Child#child{pid = NewPid},
-                            Children = lists:keyreplace(
-                                Pid, #child.pid, State#state.children, NewChild
-                            ),
-                            {ok, State#state{children = Children}}
-                    end;
-                false ->
-                    Children = lists:keydelete(Pid, #child.pid, State#state.children),
-                    {ok, State#state{children = Children}}
-            end
-    end.
-
-should_restart(_Reason, permanent) ->
-    true;
-should_restart(_Reason, temporary) ->
-    false;
-should_restart(Reason, transient) ->
-    case Reason of
-        normal -> false;
-        _any -> true
-    end.
+    %% We should always keep the start list in order for later one_for_all restarts.
+    lists:reverse(StartedC).
 
+% @hidden
 handle_call({start_child, ChildSpec}, _From, #state{children = Children} = State) ->
     Child = child_spec_to_record(ChildSpec),
     #child{id = ID} = Child,
     case lists:keyfind(ID, #child.id, State#state.children) of
         #child{pid = undefined} ->
             {reply, {error, already_present}, State};
+        #child{pid = {restarting, _Pid}} ->
+            {reply, {error, restarting}, State};
         #child{pid = Pid} ->
             {reply, {error, {already_started, Pid}}, State};
         false ->
             case try_start(Child) of
                 {ok, Pid, Result} ->
                     UpdatedChild = Child#child{pid = Pid},
-                    {reply, Result, State#state{children = [UpdatedChild | Children]}};
+                    %% The last child to start should always be at the end of the child
+                    %% start list.
+                    {reply, Result, State#state{children = Children ++ [UpdatedChild]}};
```

Comment:
Maybe we should keep the list reversed, i.e. append last item at the head in the most natural linked list way.